### PR TITLE
Issue/414

### LINKED ISSUE #414: Add a .hasNext() assertion method for Iterator<?>
I'd have thought that this method already existed but it appears that in fact, no...

It'd be a nice-to-have; along with `.doesNotHaveNext()`.

Comments?

### PR DESCRIPTION
**Pull request by stovocor**
#### Check List:
* Fixes #414 
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

File: src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterator_Test.java [hunk 1]
```diff
@@ -67,7 +67,7 @@ public void should_create_Assert() {
   @Test
   public void should_initialise_actual() {
     Iterator<String> names = asList("Luke", "Leia").iterator();
-    Iterable<String> actual = (Iterable<String>) assertThat(names).actual;
+    Iterable<String> actual = (Iterable<String>) assertThat(names).toIterable().actual;
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterator_Test.java]: this test class should only deal with `Iterator`, `actual` should be an `Iterator` and not an `Iterable`

File: src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterator_Test.java [hunk 2]
```diff
@@ -159,6 +159,7 @@ public void isNotSameAs_should_check_the_original_iterator_without_consuming_it(
     Assertions.fail("Expected assertionError, because assert notSame on same iterator.");
   }
 
+  /* TODO
   @Test
   public void iterator_can_be_asserted_twice_even_though_it_can_be_iterated_only_once() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterator_Test.java]: on the contrary, withe incoming changes, iterator cannot be asserted twice, the test should show that

File: src/test/java/org/assertj/core/api/AutoCloseableSoftAssertionsTest.java [hunk 3]
```diff
@@ -95,7 +95,7 @@ public void should_be_able_to_catch_exceptions_thrown_by_all_proxied_methods() {
       softly.assertThat(new int[] { 24 }).isEqualTo(new int[] { 25 });
 
       softly.assertThat((Iterable<String>) list("26")).isEqualTo(list("27"));
-      softly.assertThat(list("28").iterator()).contains("29");
+      softly.assertThat(list("28")).contains("29");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/AutoCloseableSoftAssertionsTest.java]: Not sure to get this change, we still want an assertion involving `Iterator`

File: src/test/java/org/assertj/core/api/BDDAssertions_then_Test.java [hunk 4]
```diff
@@ -105,7 +105,7 @@ public void then_Iterable() {
   @Test
   public void then_Iterator() {
     Iterator<String> iterator = Arrays.asList("1").iterator();
-    then(iterator).contains("1");
+    then(iterator).toIterable().contains("1");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/BDDAssertions_then_Test.java]: use `hasNext` instead of `contains` (do not use `toIterable()` we want to check `Iterator`)

File: src/test/java/org/assertj/core/api/SoftAssertionsTest.java [hunk 5]
```diff
@@ -244,7 +244,7 @@ public void should_be_able_to_catch_exceptions_thrown_by_all_proxied_methods() t
       softly.assertThat(new int[] { 24 }).isEqualTo(new int[] { 25 });
 
       softly.assertThat((Iterable<String>) Lists.newArrayList("26")).isEqualTo(Lists.newArrayList("27"));
-      softly.assertThat(Lists.newArrayList("28").iterator()).contains("29");
+      softly.assertThat(Lists.newArrayList("28")).contains("29");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/SoftAssertionsTest.java]: keep `.iterator()` but use `hasNext`

**Comment by joel-costigliola** [comment]
one recurrent comment is do not use `toIterable()` just to make the tests to pass as we want to test `Iterator` not `Iterable`. It the assertions used is not available anymore for `Iterator` use one that exists like `hasNext`

File: src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterator_Test.java (hunk 2 above)
Comment (stovocor) [reply src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterator_Test.java]: Why should it not be asserted twice? All of my new assertions do not modify the iterator at all.

File: src/test/java/org/assertj/core/api/AutoCloseableSoftAssertionsTest.java (hunk 3 above)
Comment (stovocor) [reply src/test/java/org/assertj/core/api/AutoCloseableSoftAssertionsTest.java]: I am not sure if I understand you correctly. Should I add a consuming `contains` assertion or not?

**Comment by stovocor** [comment]
I performed a diff to master after the last push today and as far as I can see, the other toIterable calls are used in classes to directly related to the Iterator assertion.

File: src/main/java/org/assertj/core/api/AbstractIteratorAssert.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * Copyright 2012-2018 the original author or authors.
+ */
+package org.assertj.core.api;
+
+import org.assertj.core.internal.Iterators;
+import org.assertj.core.util.VisibleForTesting;
+
+import java.util.Iterator;
+
+public abstract class AbstractIteratorAssert<SELF extends AbstractIteratorAssert<SELF, ELEMENT>, ELEMENT>
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractIteratorAssert.java]: add javadoc with code examples for all assertions + `toIterable()`

File: src/main/java/org/assertj/core/api/Assertions.java [hunk 7]
```diff
@@ -2635,18 +2635,18 @@ public static void useDefaultDateFormatsOnly() {
...
 
   /**
-   * Creates a new instance of <code>{@link IterableAssert}</code>.
+   * Creates a new instance of <code>{@link IteratorAssert}</code>.
    * <p>
-   * <b>Be aware that calls to most methods on returned IterableAssert will consume Iterator so it won't be possible to
-   * iterate over it again.</b> Calling multiple methods on returned IterableAssert is safe as Iterator's elements are
-   * cached by IterableAssert first time Iterator is consumed.
+   *  The {@link IteratorAssert} has limited assertions because it does not consume elements.
+   *  In order to access assertions from {@link IterableAssert}, use {@link IteratorAssert#toIterable()}.
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/Assertions.java]: let's add a code example showing `toIterable()` usage

File: src/main/java/org/assertj/core/api/IterableAssert.java [hunk 8]
```diff
@@ -217,7 +217,7 @@ public int size() {
 
   }
 
-  private static <T> Iterable<T> toIterable(Iterator<T> iterator) {
+  static <T> Iterable<T> toIterable(Iterator<T> iterator) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/IterableAssert.java]: since we have proper Iterator assertions, we can get rid of `LazyIterable` and all the places where is was used.

File: src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterator_Test.java [hunk 9]
```diff
@@ -159,120 +159,128 @@ public void isNotSameAs_should_check_the_original_iterator_without_consuming_it(
...
+    assertThat(names).hasNext().hasNext();
   }
 
-  @Test
-  public void should_pass_if_actual_and_sequence_are_equal_according_to_custom_comparison_strategy() {
-    Iterator<String> names = asList("Luke", "Leia").iterator();
-    String[] sequence = { "LUKE", "lEIA" };
-    assertThat(names).usingElementComparator(CaseInsensitiveStringComparator.instance).startsWith(sequence);
-  }
+  /*
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterator_Test.java]: let's remove that comment (I don't think anything in it applies to `Iterator`)

File: src/test/java/org/assertj/core/api/AutoCloseableSoftAssertionsTest.java (hunk 3 above)
Comment (joel-costigliola) [reply src/test/java/org/assertj/core/api/AutoCloseableSoftAssertionsTest.java]: sorry I was not clear!
what I we need here is to add an `Iterator` assertion (`hasNext` will do) to make sure they work with soft assertions.
It probably easier to add it after `softly.assertThat("48").isEqualTo("49");` to avoid changing the error messages order in the catch block.

File: src/test/java/org/assertj/core/api/SoftAssertionsTest.java [hunk 10]
```diff
@@ -244,7 +244,7 @@ public void should_be_able_to_catch_exceptions_thrown_by_all_proxied_methods() t
...
       softly.assertThat(Lists.newArrayList("30")).isEqualTo(Lists.newArrayList("31"));
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/SoftAssertionsTest.java]: add an Iterator assertion (hasNext will do) to make sure they work with soft assertions (easier to add it just before the catch block to avoid changing the error messages order afterwards.

File: src/test/java/org/assertj/core/api/SoftAssertionsTest.java [hunk 11]
```diff
@@ -589,6 +589,7 @@ public void should_pass_when_using_extracting_with_iterator() {
 
     try (AutoCloseableSoftAssertions softly = new AutoCloseableSoftAssertions()) {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/SoftAssertionsTest.java]: remove this test as Iterator don't support extracting

File: src/test/java/org/assertj/core/api/WithAssertions_delegation_Test.java [hunk 12]
```diff
@@ -393,7 +393,7 @@ public void withAssertions_assertThat_iterable_assert_factory_Test() {
    */
   @Test
   public void withAssertions_assertThat_iterator_Test() {
-    assertThat(Arrays.asList(ITEMS).iterator()).contains(ITEMS[0]);
+    assertThat(Arrays.asList(ITEMS).iterator()).toIterable().contains(ITEMS[0]);
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/WithAssertions_delegation_Test.java]: should br an iterator assertion, e.g:
```
assertThat(Arrays.asList(ITEMS).iterator()).hasNext();
```

File: src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_Test.java [hunk 13]
```diff
@@ -81,7 +81,7 @@ public void should_fail_if_filter_is_on_private_field_and_reading_private_field_
 
   @Test
   public void should_filter_iterator_under_test_on_property_values() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_Test.java]: remove, `filter` is not supported for Iterator

File: src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_condition_Test.java [hunk 14]
```diff
@@ -61,7 +61,7 @@ public void should_fail_if_given_condition_is_null() {
   }
 
   @Test
-  public void shoul_honor_AssertFactory_strongly_typed_navigation_assertions() {
+  public void should_honor_AssertFactory_strongly_typed_navigation_assertions() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_condition_Test.java]: good catch!

File: src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_in_Test.java [hunk 15]
```diff
@@ -63,7 +63,7 @@ public void should_fail_if_filter_is_on_private_field_and_reading_private_field_
 
   @Test
   public void should_filter_iterator_under_test_on_property_values() {
-    assertThat(employees.iterator()).filteredOn("age", in(800)).containsOnly(yoda, obiwan);
+    assertThat(employees.iterator()).toIterable().filteredOn("age", in(800)).containsOnly(yoda, obiwan);
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_in_Test.java]: remove (not supported for Iterator)

File: src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_notIn_Test.java [hunk 16]
```diff
@@ -62,7 +62,7 @@ public void should_fail_if_filter_is_on_private_field_and_reading_private_field_
 
   @Test
   public void should_filter_iterator_under_test_on_property_values() {
-    assertThat(employees.iterator()).filteredOn("age", notIn(800)).containsOnly(luke, noname);
+    assertThat(employees.iterator()).toIterable().filteredOn("age", notIn(800)).containsOnly(luke, noname);
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_notIn_Test.java]: remove test as it does not apply to Iterator assertions

File: src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_not_Test.java [hunk 17]
```diff
@@ -66,7 +66,7 @@ public void should_fail_if_filter_is_on_private_field_and_reading_private_field_
 
   @Test
   public void should_filter_iterator_under_test_on_property_values() {
-    assertThat(employees.iterator()).filteredOn("age", not(800)).containsOnly(luke, noname);
+    assertThat(employees.iterator()).toIterable().filteredOn("age", not(800)).containsOnly(luke, noname);
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/iterable/IterableAssert_filteredOn_not_Test.java]: remove test as it does not apply to Iterator assertions

File: src/test/java/org/assertj/core/api/iterator/IteratorAssert_hasNext_Test.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+  @Override
+  protected IteratorAssert<Object> invoke_api_method() {
+    return assertions.hasNext();
+  }
+
+  @Override
+  protected void verify_internal_effects() {
+    verify(iterators).assertHasNext(getInfo(assertions), getActual(assertions));
+  }
+
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/iterator/IteratorAssert_hasNext_Test.java]: add test to show that it fails when actual is null, something like:
```
  @Test
  public void should_fail_if_actual_is_null() {
    assertThatExceptionOfType(AssertionError.class).isThrownBy(() -> assertThat(iterator).hasNext())))
                                                   .withMessage(FailureMessages.actualIsNull());
  }
```

File: src/test/java/org/assertj/core/api/iterator/IteratorAssert_isExhausted_Test.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+  @Test
+  public void should_pass_for_exhausted_iterator() {
+    Iterator<Object> iterator = emptyList().iterator();
+    assertThat(iterator).isExhausted();
+  }
+
+  @Test
+  public void should_fail_if_actual_has_at_least_one_element() {
+    // GIVEN
+    SoftAssertions softly = new SoftAssertions();
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/iterator/IteratorAssert_isExhausted_Test.java]: use standard assertions (no need of soft ones)

File: src/test/java/org/assertj/core/perf/SoftAssertionsPerfTest.java [hunk 20]
```diff
@@ -216,7 +216,7 @@ public void should_be_able_to_catch_exceptions_thrown_by_all_proxied_methods() t
     softly.assertThat(22).isEqualTo(23);
     softly.assertThat(new int[] { 24 }).isEqualTo(new int[] { 25 });
     softly.assertThat((Iterable<String>) Lists.newArrayList("26")).isEqualTo(Lists.newArrayList("27"));
-    softly.assertThat(Lists.newArrayList("28").iterator()).contains("29");
+    softly.assertThat(Lists.newArrayList("28").iterator()).toIterable().contains("29");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/perf/SoftAssertionsPerfTest.java]: use `hasNext` instead of `.toIterable()` + `contains`

**Comment by joel-costigliola** [comment]
Integrated thanks @stovocor!
